academy/xl-run-prod/6-migration.md

```diff
@@ -0,0 +1,81 @@
+---
+title: Prepare and do migrations
+order: 6
+description: Keep your software and network up to date
+tag: deep-dive
+---
+
+# Prepare and do migrations
+
+Your software is now running, that is good. Time passes and a software upgrade vote proposal passed. You need to upgrade your node to the next version. The state may need to migrate too. When the state needs to migrate, it is also said that there are _breaking changes_. This state migration is the upgrade step that takes most of the time.
+
+You can do the whole process somewhat manually or use a tool to assist you do it smoothly and fast. This is not an unreasonable concern since, when you upgrade, all nodes are simultaneously down.
+
+The main tool is [Cosmovisor](https://docs.cosmos.network/main/run-node/cosmovisor.html). It is a wrapper executable that in turn launches your node as a subprocess. It watches our for when the node stops and drops upgrade information.
+
+Before you upgrade your mainnet, it is good practice to upgrade your testnet(s). You can also use your mainnet state in a temporary testnet to test the computation needs of your state upgrade.
+
+<!-- TODO at which versions does the below apply? Anyway we want to use the latest advised way -->
+
+## Set up Cosmovisor
+
+Cosmovisor is a piece of software that you need to [install](https://docs.cosmos.network/main/run-node/cosmovisor.html#installation) on your node computer. Its configuration is done via:
+
+1. [Environment variables](https://docs.cosmos.network/main/run-node/cosmovisor.html#command-line-arguments-and-environment-variables).
+2. [Configuration files and folders](https://docs.cosmos.network/main/run-node/cosmovisor.html#folder-layout) that are polled at interval and which you prepare by hand or via [a command](https://dodocs.cosmos.network/mainun-node/cosmovisor.html#initialization).
+
+When starting you can pass it command-line arguments that it will pass on to the wrapped Cosmos chain executable. Typically, `cosmovisor run start`, where `start` is the same as in `myprojectd start`.
+
+What Cosmovisor does when it launches is:
+
+1. Start and launch your node executable, say `myprojectd`.
+2. Poll regularly, for instance in `.myprojectd/data/upgrade-info.json`, for potential upgrade information. Note how this file is in the `data` folder and created by the [`x/upgrade` module](https://docs.cosmos.network/main/building-modules/upgrade.html) when appropriate.
+3. When an upgrade information is available, it will wait for the executable to stop on its own, in effect after it has committed all state at the given block height.
```

Comment:
Step 3 is a little tricky:

_When **an upgrade information** is available..._
> either "When upgrade information" OR "When an upgrade information _something_" 

_...after it has committed **all state** at the given block height._
> "all state**s**"?

Reply:
ok, I will adjust myself.